### DISCUSSION TITLE
Does Cozo support pagination? Keyset pagination?

### DISCUSSION THREAD

**Discussion by andrewbaxter** [cozodb/cozo#209] [Q&A] [answered]
AFAICT looking through the rust docs it returns the whole query set. I imagine you could implement something similar yourself, but you'd need to use a new query to do so (possibly multiple queries depending on where you left off).

**Comment by aramallo** [comment]
@andrewbaxter you can use the `:limit` and `:offset` options to paginate yourself.

**Reply by andrewbaxter** [reply, marked as answer]
Oh thanks, yeah! I actually just saw that the other day.

I was specifically interested in keyset pagination to avoid missed/duplicate results when fetching large sets of data while the data is changing.

There _is_ result ordering too though.  So I guess you could probably order it then grab the key from the last entry and make the next query with `> key`? I assume it fetches the whole result set on the server before doing that, but at least maybe not all of it would go over the wire.

**Reply by aramallo** [reply]
Yes I think it should work and if you are using Rockdb storage, you might be able to leverage the fact that keys are lexicographically sorted to make the server side more efficient.
